### PR TITLE
Migrate to 7TV v3 api

### LINKED ISSUE #410: Migrate 7TV API endpoints
### Checklist

- [X] I have checked the [issue page](https://github.com/lay295/TwitchDownloader/issues) for duplicates
- [X] I am running the latest version ([download here](https://github.com/lay295/TwitchDownloader/releases/latest))

### Write your feature request here

As I've stated previously [here,](https://github.com/lay295/TwitchDownloader/pull/337#issuecomment-1290973132) the 7TV endpoints should be migrated to the v3 of the API. v2 is still up, but will most likely get completely disabled soon. I tried cloning the repo and doing it myself, but unfortunately, couldn't finish it.

The global emote set endpoint would be:
https://7tv.io/v3/emote-sets/global

The user emote set endpoint would be, for example:
https://7tv.io/v3/users/twitch/70647828 (for my channel)

### PR DESCRIPTION
**Pull request by ScrubN**
Fixes #410

### CODE REVIEW

**Comment by sunkhaskasis** [comment]
I just tried it on my end and it didn't work, throws the `ERROR: Error reading JArray from JsonReader. Current JsonReader item is not an array: StartObject. Path '', line 1, position 1.` error, though I'm not sure if that's what causing it.

**Comment by ScrubN** [comment]
The one time I don't test...

Working on it.

**Comment by ScrubN** [comment]
Found the problem! The old api was structured:
```
[
    {
    },
    {
    },
]
```
where each `{}` is an emote. The new api is structured:
```
{
    ...
    "emotes": [
    {
    },
    {
    }
    ]
    ....
}
```
where each `{}` is an emote.

The key difference is that the new API isn't wrapped in `[]` making it an invalid JSON. Thanks 7tv!

**Comment by ScrubN** [comment]
The old and new APIs looked close enough to me when I initially checked their outputs, but I should've looked further. If this hadn't happened then it would've failed in other ways. They completely restructured the API, it wasn't just a renaming like last time.

**Comment by goldbattle** [comment]
It should still be valid (its an array of json objects).
Should be able to first create a JSON object and get its value for the
"emotes" which is then the json array you expect.

**Comment by ScrubN** [comment]
> It should still be valid (its an array of json objects). Should be able to first create a JSON object and get its value for the "emotes" which is then the json array you expect.

Yes, that was more or less the plan. But many properties were renamed or moved, for example the `mime` field is gone and has been replaced by `"animated" : <bool>` and an array of valid formats in a variety of sizes.

**Comment by goldbattle** [comment]
To get the file you need to get it from the "host"

- `host > url`: cdn.7tv.app/emote/603caa69faf3a00014dff0b1
- `host > files > name`: [scale].webp
- Gives https://cdn.7tv.app/emote/603caa69faf3a00014dff0b1/1x.webp

You might be able to just assume there is always there is a webp there.
For example chatterino [only uses webp](https://github.com/Chatterino/chatterino2/blob/a033dbc933ed87fc0d1ca4497c14401b9e16c423/src/providers/seventv/SeventvEmotes.cpp#L96) versions currently.

**Comment by ScrubN** [comment]
Really annoying thing, zero width is now stored as a flag and I'm having a hard time finding their definition file for the emote flags.

**Comment by goldbattle** [comment]
Here are the definitions:
https://github.com/SevenTV/API/blob/dev/data/model/emote.model.go#L60-L71
https://github.com/Chatterino/chatterino2/blob/dd6cb80ab945a4f0a40da9da8de83eea2de1ce08/src/providers/seventv/SeventvEmotes.hpp#L27-L47

There are actually a few extra checks that need to be made that is not currently checked.
1. Emote is not an unlisted one (these shouldn't be seen by users), can be enforced by checking `"listed": true,` in json
2. Emote has not been disallowed on twitch, this would need to be checked in the flags `1 << 24`

If you want good examples see sodapoppin's channel:
- https://7tv.io/v3/users/twitch/26301881
- `WICKEDglasses` has flags 256 (zero width is 2^8 (eighth bit is 1))
- `ShowerTime` has flags 256 (zero width is 2^8 (eighth bit is 1))

**Comment by ScrubN** [comment]
My adaptation is complete. If you test it for yourself, odds are you'll end up with a slightly smaller output json because the streamer has unlisted emotes. I used Mizkif and he has an unlisted emote that the v2 parser did not ignore.

Please let me know if you have any critiques of my code, this is my first time properly working with Newtonsoft.Json

File: TwitchDownloaderCore/TwitchHelper.cs [hunk 1]
```diff
@@ -162,39 +164,67 @@ public static async Task<EmoteResponse> GetThirdPartyEmoteData(string streamerId
...
-                    string name = emote["name"].ToString();
-                    string mime = emote["mime"].ToString().Split('/')[1];
-                    string url = String.Format("https://cdn.7tv.app/emote/{0}/[scale]x", id);
-                    EmoteResponseItem emoteResponse = new EmoteResponseItem() { Id = id, Code = name, ImageType = mime, ImageUrl = url };
-                    if (emote["visibility_simple"].ToList().Contains("ZERO_WIDTH"))
+                    string emoteId = stvEmote["id"].ToString();
+                    string emoteName = stvEmote["name"].ToString();
+                    JObject emoteData = (JObject)stvEmote["data"];
+                    JObject emoteHost = (JObject)emoteData["host"];
+                    JArray emoteFiles = JArray.Parse(emoteHost["files"].ToString());
```
Comment (goldbattle) [inline TwitchDownloaderCore/TwitchHelper.cs:196]: Probably don't need to convert back, probably just directly cast it to (JArray) as you have for JObject already.
https://www.newtonsoft.com/json/help/html/QueryJson.htm

**Comment by sunkhaskasis** [comment]
Does SkiaSharp not support AVIF? IMO, it'd be better if it was preferred
instead of WEBP, as it features better compression and weighs less.

File: TwitchDownloaderCore/TwitchHelper.cs (hunk 1 above)
Comment (ScrubN) [reply TwitchDownloaderCore/TwitchHelper.cs:196]: When I use a `JArray` cast, the breakpoint debugger exits step mode when it hits that line, so I'd rather keep it as `JArray.Parse` so breakpoints still work after the statement

**Comment by ScrubN** [comment]
> Does SkiaSharp not support AVIF? IMO, it'd be better if it was preferred instead of WEBP, as it features better compression and weighs less.

Looks like SkiaSharp does support AVIF. I just figured may as well prefer webp since Chatterino does

File: TwitchDownloaderCore/TwitchHelper.cs [hunk 2]
```diff
@@ -162,39 +164,67 @@ public static async Task<EmoteResponse> GetThirdPartyEmoteData(string streamerId
...
+                JArray stvEmotes = JArray.Parse(globalEmoteElement.ToString());
 
                 if (streamerId != null)
                 {
                     //Channel might not have 7TV emotes
                     try
                     {
-                        STV.Merge(JArray.Parse(await httpClient.GetStringAsync(String.Format("https://api.7tv.app/v2/users/{0}/emotes", streamerId))));
+                        string streamerEmoteResponse = await httpClient.GetStringAsync(string.Format("https://7tv.io/v3/users/twitch/{0}", streamerId));
+                        JsonElement streamerEmoteSetElement = JsonDocument.Parse(streamerEmoteResponse).RootElement.GetProperty("emote_set");
```
Comment (goldbattle) [inline TwitchDownloaderCore/TwitchHelper.cs:183]: I would just use JObject here if you can (this JsonElement is a different json library I believe? might be wrong).
Don't know if this compiles, but I would try this:
```
JObject stvChannel = JObject.Parse(await httpClient.GetStringAsync("....." + streamerId));
stvEmotes.Merge((JArray)stvChannel["emote_set"]["emotes"]);
```

**Comment by ScrubN** [comment]
Also the project currently is on SkiaSharp 2.80.2. In #407 I updated the SkiaSharp packages to 2.88.3.

File: TwitchDownloaderCore/TwitchHelper.cs (hunk 2 above)
Comment (ScrubN) [reply TwitchDownloaderCore/TwitchHelper.cs:183]: Yep that works. Also using that syntax to simplify the global response.

**Pull request merged**